crates/matrix-sdk/src/event_cache/deduplicator.rs

```diff
@@ -0,0 +1,266 @@
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Simple but efficient types to find duplicated events. See [`Deduplicator`]
+//! to learn more.
+
+use std::{collections::BTreeSet, sync::Mutex};
+
+use growable_bloom_filter::{GrowableBloom, GrowableBloomBuilder};
+
+use super::store::{Event, RoomEvents};
+
+/// `Deduplicator` is an efficient type to find duplicated events.
+///
+/// It uses a [bloom filter] to provide a fast probabilistic answer to: “has
+/// event E been seen already?”. False positives are possible, while false
+/// negatives are impossible. In the case of a positive reply, we fallback to a
+/// linear (backward) search on all events to check whether it's a false
+/// positive or not
+///
+/// [bloom filter]: https://en.wikipedia.org/wiki/Bloom_filter
+pub struct Deduplicator {
+    bloom_filter: Mutex<GrowableBloom>,
+}
+
+impl Deduplicator {
+    const APPROXIMATED_MAXIMUM_NUMBER_OF_EVENTS: usize = 800_000;
+    const DESIRED_FALSE_POSITIVE_RATE: f64 = 0.001;
+
+    /// Create a new `Deduplicator`.
+    pub fn new() -> Self {
+        Self {
+            bloom_filter: Mutex::new(
+                GrowableBloomBuilder::new()
```

Comment:
Sorry to be the knight who says [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it), but… is it so much faster, considering we do have a btreeset check in both branches after the `check_and_set` below (equivalent to inserting into the btreeset)? Do you have data supporting this (a benchmark)?

Reply:
The `BTreeSet` is used only to check the duplicates in the _new_ events that we are checking. The bloom filter is to check the duplicates in _past_ events. For example:

* The bloom filter contains `[]`. I check `["$ev0", "$ev1", "$ev2]`. No duplicates.
* The bloom filter now contains `["$ev0", "$ev1", "$ev2"]`. I check `["$ev3"]`. No duplicates.
* The bloom filter now contains `["$ev0", "$ev1", "$ev2", "$ev3"]`. I check `["$ev0"]`. There is a duplicate.
* The bloom filter now contains `["$ev0", "$ev1", "$ev2", "$ev3"]`. I check `["$ev4", "$ev4"]`. There is a duplicate but in the _new_ events! That's detected by the use of `BTreeSet`.
* The bloom filter now contains `["$ev0", "$ev1", "$ev2", "$ev3", "$ev4"]` now.

This `BTreeSet` is necessary to efficiently detect duplicates in the _new_ scanned events. If we rely on the Bloom filter only, we cannot check the false positives, because the _new_ events are not yet added into the `RoomEvents`.

You're right the `BTreeSet` is hit in the 2 paths inside `check_and_set` but the bloom filter is still more memory-space efficient than if we were using `BTreeSet` for _new_ (currently scanned) **and** _past_ (already scanned) events. That's the only reason I use a bloom filter here: to save memory.